src/test/java/cloud/fogbow/ras/core/plugins/interoperability/cloudstack/publicip/v4_9/CloudStackPublicIpPluginTest.java

```diff
@@ -0,0 +1,798 @@
+        SuccessfulAssociateIpAddressResponse.IpAddress ipAddress =
+                Mockito.mock(SuccessfulAssociateIpAddressResponse.IpAddress.class);
+        Mockito.when(ipAddress.getId()).thenReturn(ipAddressIdExpected);
+        Mockito.when(ipAddress.getIpAddress()).thenReturn(ipExpected);
+
+        Mockito.when(response.getIpAddress()).thenReturn(ipAddress);
+        AsyncRequestInstanceState asyncRequestInstanceState = new AsyncRequestInstanceState(null, null , null);
+        String createFirewallRuleJobId = "jobId";
+
+        // verify before
+        Assert.assertNull(asyncRequestInstanceState.getIpInstanceId());
+        Assert.assertNull(asyncRequestInstanceState.getIp());
+        Assert.assertNull(asyncRequestInstanceState.getState());
+        Assert.assertNull(asyncRequestInstanceState.getCurrentJobId());
+
+        // exercise
+        this.plugin.setAsyncRequestInstanceSecondStep(response, asyncRequestInstanceState, createFirewallRuleJobId);
+
+        // verify after
+        Assert.assertEquals(ipAddressIdExpected, asyncRequestInstanceState.getIpInstanceId());
+        Assert.assertEquals(ipExpected, asyncRequestInstanceState.getIp());
+        Assert.assertEquals(AsyncRequestInstanceState.StateType.CREATING_FIREWALL_RULE,
+                asyncRequestInstanceState.getState());
+        Assert.assertEquals(createFirewallRuleJobId, asyncRequestInstanceState.getCurrentJobId());
+    }
+
+    // test case: When calling the requestCreateFirewallRule method and occurs a HttpResponseException,
+    // it must verify if It throws FogbowException.
+    @Test
+    public void testRequestCreateFirewallRuleFail() throws FogbowException, HttpResponseException {
+        // set up
+        CreateFirewallRuleRequest request = new CreateFirewallRuleRequest.Builder().build("");
+
+        PowerMockito.mockStatic(CloudStackCloudUtils.class);
+        PowerMockito.when(CloudStackCloudUtils.doRequest(Mockito.any(), Mockito.any(), Mockito.any())).
+                thenThrow(CloudstackTestUtils.createBadRequestHttpResponse());
+
+        // verify
+        this.expectedException.expect(FogbowException.class);
+        this.expectedException.expectMessage(CloudstackTestUtils.BAD_REQUEST_MSG);
+
+        // exercise
+        this.plugin.requestCreateFirewallRule(request, this.cloudStackUser);
+    }
+
+    // test case: When calling the requestCreateFirewallRule method, it must verify if It
+    // returns the right jobId.
+    @Test
+    public void testRequestCreateFirewallRuleSuccessfully() throws FogbowException, HttpResponseException {
+        // set up
+        CreateFirewallRuleRequest request = new CreateFirewallRuleRequest.Builder().build("");
+
+        String jsonResponse = "";
+        PowerMockito.mockStatic(CloudStackCloudUtils.class);
+        PowerMockito.when(CloudStackCloudUtils.doRequest(
+                Mockito.any(), Mockito.any(), Mockito.any())).thenReturn(jsonResponse);
+
+        String jobIdExpexted = "jobId";
```

Comment:
You can migrate this or any other string to a constant in the **TestUtils** class.

Reply:
I have some questions about it. I'll talk to you in the hangout.